File: matlab/getSurface.m
% This file is part of the FMI IMPEx tools.
%
% Copyright 2014- Finnish Meteorological Institute
%
% This program is free software: you can redistribute it and/or modify
% it under the terms of the GNU General Public License as published by
% the Free Software Foundation, either version 3 of the License, or
% (at your option) any later version.
%
% This program is distributed in the hope that it will be useful,
% but WITHOUT ANY WARRANTY; without even the implied warranty of
% MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
% GNU General Public License for more details.
%
% You should have received a copy of the GNU General Public License
% along with this program.  If not, see <http://www.gnu.org/licenses/>.

function filestr = getSurface( resourceID, planeNormal, planePoint, fileName, varargin )
% GETSURFACE  Get data from FMI IMPEx server on a plane grid.
%
% Required parameters:
% resourceID -- Impex ID of the desired simulation run. Use
%       getMostRelevantRun to obtain this.
% planeNormal -- a vector perpendicular to the desired plane
% planePoint -- coordinates of a point in the desired plane
% fileName -- The data is saved in this file in your working directory.
%
% Optional Name,Value pairs:
% 'variables' -- list of variable names to be printed in the output file
%       (space separated char array), default: all.
% 'resolution' -- distance between data points on the plane. The default is
%       the grid cell size used in the simulation.
% 'interpolationMethod' -- either 'linear' or 'nearestGridPoint'.
% 'outputFileType' -- either 'VOTable' or 'netCDF'.
%
% See also: GETMOSTRELEVANTRUN, GETDATAPOINTVALUE

% Written by Tiera Laitinen, Finnish Meteorological Institute, 2014.

%=========================================================================

% Parse and check input arguments
args = inputParser;
addRequired( args, 'resourceID', @ischar );
addRequired( args, 'planeNormal', @isnumeric );
addRequired( args, 'planePoint', @isnumeric );
addRequired( args, 'fileName', @ischar );
addParameter( args, 'variables', '', @ischar );
addParameter( args, 'resolution', -1, @isnumeric );
addParameter( args, 'interpolationMethod', '', @ischar );
addParameter( args, 'outputFileType', '', @ischar );
parse( args, resourceID, planeNormal, planePoint, fileName, varargin{:} );

k = size(planeNormal);
if k(1) > 1
    planeNormal = planeNormal';     % make them row vectors
end
k = size(planePoint);
if k(1) > 1
    planePoint = planePoint';
end
if length(planeNormal) ~= 3 || length(planePoint) ~= 3
    error('planeNormal and planePoint must be vectors of length 3.');
end
pituus = sqrt( planeNormal(1)^2 + planeNormal(2)^2 + planeNormal(3)^2 );
planeNormal = planeNormal / pituus;

% Compose parameter data structure
data.val.resourceID.name = 'ResourceID';
data.val.resourceID.val = resourceID;
data.val.resourceID.type = '{http://impex-fp7.oeaw.ac.at}ResourceID';
data.val.normal.name = 'PlaneNormalVector';
data.val.normal.val = num2str( planeNormal );
data.val.point.name = 'PlanePoint';
data.val.point.val = num2str( planePoint );
if ~ strcmp( args.Results.variables, '' )
    data.val.variable.name = 'Variable';
    data.val.variable.val = args.Results.variables;
    data.val.variable.type = '{http://impex-fp7.fmi.fi}Variable';
end
if args.Results.resolution > 0
    data.val.extra.name = 'extraParams';
    data.val.extra.val.resolution.name = 'Resolution';
    data.val.extra.val.resolution.val = args.Results.resolution;
end
if ~ strcmp( args.Results.interpolationMethod, '' )
    data.val.extra.name = 'extraParams';
    data.val.extra.val.interpolationMethod.name = 'InterpolationMethod';
    data.val.extra.val.interpolationMethod.val = args.Results.interpolationMethod;
    data.val.extra.val.interpolationMethod.type = '{http://impex-fp7.fmi.fi}enumInterpolation';
end
if ~ strcmp( args.Results.outputFileType, '' )
    data.val.extra.name = 'extraParams';
    data.val.extra.val.outputFileType.name = 'OutputFileType';
    data.val.extra.val.outputFileType.val = args.Results.outputFileType;
    data.val.extra.val.outputFileType.type = '{http://impex-fp7.fmi.fi}OutputFormat';
end

% Call the SOAP service
soapMessage = createSoapMessage( ...
    'http://impex-fp7.fmi.fi', ...
    'getSurface', ...
    data, 'document');
response = callSoapService( ...
    'http://impex-fp7.fmi.fi/ws/Methods_FMI.php', ...
    'getSurface', ...
    soapMessage);

% Parse the SOAP response, get data file and save
i1 = strfind( response, '<ns1:anyURI>' ) + 12;
i2 = strfind( response, '</ns1:anyURI>' ) - 1;
urlAns = response(i1(1) : i2(1));

if strncmp( urlAns, 'http', 4 )
    [filestr, fileStatus] = urlwrite( urlAns, fileName );
else    
    display('SOAP error: ');
    display( urlAns );
    return;
end

if ~ fileStatus
    display( urlAns );
    error('Could not save the data file from the address above');
end

return;

File: web/functions/getSurface.php
<?php

// This file is part of the FMI IMPEx tools.
//
// Copyright 2014- Finnish Meteorological Institute
//
// This program is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program.  If not, see <http://www.gnu.org/licenses/>.
// #####################################################################################################
//
// 								      		getSurface.php
//
// #####################################################################################################
// =====================================================================================================
// Compute interpolated values on a plane surface for parameters defined in $params input variable.
//
//	$params = array
//		string 'ResourceID'		   x  Resource ID of the <NumericalOutput> element in tree.xml
//		string 'Variable'		  	  List of individual parameters from the selected
//									  NumericalOutput separated by a comma.
//									  Identified by Spase/NumericalOutput/Parameter/ParameterKey.
//									  By default: all parameters in the NumericalOutput are sent back.
//		string 'PlaneNormalVector' x  3D components of a normalized vector normal to the requested plane
//                                    = string of 3 floats delimited by " " or ","
//		string 'PlanePoint'        x  A 3D coordinate indicating the position of one point in the
//                                    requested plane.
//                                    = string of 3 floats delimited by " " or ","
//		array 'extraParams'			  Associative array of additional parameters. Different for each SMDB
//			float 'Resolution'	          Spatial resolution of the mesh grid (The unit defined in
//										  <SimulationRun>)
//                                        Default : basic grid size
//			string 'InterpolationMethod'  Enumerated list of interpolation method to be defined
//								          (NearestGridPoint, Linear (default))
//			string 'OutputFiletype'	      Enumerated list of output file formats.
//									      FMI supports ASCII, VOTable (default) and netCDF.
//
//	Parameters indicated by 'x' are mandatory.
// =====================================================================================================

// =====================================================================================================
// Version 2.0 2015-06-01		First release of the new code.
// =====================================================================================================

function getSurface($params) {

	// --------------------------------
	// Log caller info and method name
	// --------------------------------

	log_method("getSurface", $params);


	// ----------------------------------------------------------------------------
	// Define a structure to handle Input parameters.
	// The user given values will be stored in the 'data' field.
	// The 'data' field is initially set to default value.
	// ----------------------------------------------------------------------------

	$InputParams = [
		'ResourceID'  		=> ['type' => 'string',   'mandatory' => true,  'data' => null],
		'Variable'    		=> ['type' => 'string',   'mandatory' => false, 'data' => ""],
		'PlaneNormalVector' => ['type' => 'vector',   'mandatory' => true,  'data' => null],
		'PlanePoint' 		=> ['type' => 'vector',   'mandatory' => true,  'data' => null],
		'extraParams' => [
			'Resolution' 		  => ['type' => 'float',  'mandatory' => false, 'data' => null],
			'InterpolationMethod' => ['type' => 'string', 'mandatory' => false, 'enum' => ['Linear', 'NearestGridPoint'], 'data' => 'Linear'],
			'OutputFileType'      => ['type' => 'string', 'mandatory' => false, 'enum' => ['netCDF', 'VOTable', 'ASCII'], 'data' => 'VOTable']
		]
	];


	// --------------------------------------------------
	// Handle the user given input parameters ($params).
	// --------------------------------------------------

	// Check that mandatory parameters are included in user given input parameter
	// values ($params) and copy the values of all parameters to $InputParams.
	// Check also that given parameters are of proper type.
	// Set also default values to parameters that the user has not defined.
	$InputParams = check_input_params($InputParams, $params);

	// Check the ResourceID and convert to new format if it is in old format (FMI only).
	$InputParams['ResourceID']['data'] = check_resourceID($InputParams['ResourceID']['data']);

	// Load the tree file and get the NumericalOutput and SimulationRun elements
	$tree_xml = get_Tree($InputParams['ResourceID']['data']);
	$num_output_elem = get_NumericalOutput($tree_xml, $InputParams['ResourceID']['data']);
	$simulation_run_elem = get_SimulationRun($tree_xml, $num_output_elem);

	// Read the user defined list of variables into $InputParams['Variable']['data']
	// as an array (key = variable name, value = variable description) and check
	// that $num_output_elem contains the given variable names.
	$InputParams['Variable']['data'] = get_variable_list($InputParams['Variable']['data'], $num_output_elem);

	// Read some metadata from $tree_xml into $metadata global variable
	get_metadata($tree_xml, $simulation_run_elem, $num_output_elem, $InputParams);

	// Normalize 'PlaneNormalVector' to unit vector
	$InputParams['PlaneNormalVector']['data'] = unit_vector($InputParams['PlaneNormalVector']['data']);

	// Check the input parameter PlaneNormalVector. In general case it could be pointing to any direction.
	// However, here we accept only cases where PlaneNormalVector is parallel to any three coordinate axis.
	// The general case remains to be written.
	// Vector which is parallel to any of the three coordinate axis will have two zero components.
	$v = $InputParams['PlaneNormalVector']['data'];
	if 	(($v[0] != 0.0) and ($v[1] != 0.0)) throw_error('Client', ERROR_NORMAL_VECTOR_DIR);
	if 	(($v[0] != 0.0) and ($v[2] != 0.0)) throw_error('Client', ERROR_NORMAL_VECTOR_DIR);
	if 	(($v[1] != 0.0) and ($v[2] != 0.0)) throw_error('Client', ERROR_NORMAL_VECTOR_DIR);

	// If Resolution not defined then set to default value
	if ($InputParams['extraParams']['Resolution']['data'] == null) {
		$InputParams['extraParams']['Resolution']['data'] = get_basic_grid_cell_size($simulation_run_elem);
	}


	// -----------------------------------------------------------------------------
	// Generate the plane mesh and store the points in a simple ascii file which is
	// readable by hcintpol.
	// -----------------------------------------------------------------------------

	if ($v[2] != 0.0) 	// XY plane
		$point_file_name = write_plane_mesh($simulation_run_elem, $InputParams['PlanePoint']['data'], "XY", $InputParams['extraParams']['Resolution']['data']);
	if ($v[0] != 0.0)	// YZ plane
		$point_file_name = write_plane_mesh($simulation_run_elem, $InputParams['PlanePoint']['data'], "YZ", $InputParams['extraParams']['Resolution']['data']);
	if ($v[1] != 0.0)	// XZ plane
		$point_file_name = write_plane_mesh($simulation_run_elem, $InputParams['PlanePoint']['data'], "XZ", $InputParams['extraParams']['Resolution']['data']);


	// --------------------------------------------------
	// Handle the computation of interpolated values.
	// --------------------------------------------------

	$final_file_name = handle_interpolation($point_file_name, $num_output_elem, $simulation_run_elem, $InputParams);


	// --------------
	// Final touches
	// --------------

	clean_up([$point_file_name]);
	return(returnURL($final_file_name));

}	// THE END
?>
